packages/terra-folder-tree/src/FolderTree.jsx

```diff
@@ -54,6 +55,72 @@ const FolderTree = ({
   intl,
 }) => {
   const folderTreeID = `folder-tree-${uuidv4()}`;
+  const listNode = useRef();
+
+  const handleKeyDown = event => {
+    const listItems = listNode.current.querySelectorAll('[data-item-show-focus=true]');
+    // Remove all hidden list items
+    const visibleListItems = Array.prototype.slice.call(listItems).filter((item) => {
+      let parent = item.parentNode;
+      while (parent && parent !== listNode.current) {
+        if (parent.hasAttribute('hidden')) {
+          return false;
+        }
+        parent = parent.parentNode;
+      }
+      return true;
+    });
+    const currentIndex = Array.from(visibleListItems).indexOf(event.target);
+    const lastIndex = visibleListItems.length - 1;
+
+    const handleHomeKey = () => visibleListItems[0].focus();
+    const handleEndKey = () => visibleListItems[lastIndex].focus();
+
+    switch (event.nativeEvent.keyCode) {
+      case KeyCode.KEY_END:
+        event.preventDefault();
+        handleEndKey();
+        break;
+      case KeyCode.KEY_HOME:
+        event.preventDefault();
+        handleHomeKey();
+        break;
+      case KeyCode.KEY_UP: {
+        event.preventDefault();
+        const previousIndex = currentIndex - 1;
+        visibleListItems[previousIndex]?.focus();
+        break;
+      }
+      case KeyCode.KEY_DOWN: {
+        event.preventDefault();
+        const nextIndex = currentIndex + 1;
+        visibleListItems[nextIndex]?.focus();
+        break;
+      }
```

Comment:
Shouldn't we be able to use previousElementSibling and nextElementSibling to simplify the logic?

Reply:
These can't be used since subFolders are grouped as `<ul>`s. And subFolders and items can be sibling with each other